tests/metrics/annotationUtil.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"sync"
+)
+
+//Client request grafana API on a set of resource paths.
+type client struct {
+	// base is the root URL for all invocations of the client
+	baseUrl url.URL
+	client  *http.Client
+}
+
+//Annotation is a specification of the desired behavior of adding annotation
+type Annotation struct {
+	DashboardId         int
+	PanelId             int
+	TimestampInMilliSec int64
+	Tags                []string
+	Text                string
+}
+
+//NewClient creats a new grafanaClient. This client performs rest functions
+//such as Get, Post on specified paths.
+func NewClient(grafanaUrl string, userName string, password string, prometheusExporterPort int) (*client, error) {
+	u, err := url.Parse(grafanaUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	initFunc(prometheusExporterPort)
+	u.User = url.UserPassword(userName, password)
+	return &client{
+		baseUrl: *u,
+		client:  &http.Client{},
+	}, nil
+}
+
+func (annotation *Annotation) getBody() ([]byte, error) {
+	m := map[string]interface{}{
```

Comment:
consider using struct annotation (`json:"field_name"`) to avoid creating an extra map?

Reply:
Good idea and accept